routing/router.go

```diff
@@ -2293,3 +2293,265 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+
+	// In the first pass, build a complete list of route pubkeys and fetch
+	// policies for hops that are specified by channel id.
+
+	// Initialize the route pubkeys list with the source node.
+	pubkeys := []route.Vertex{source}
+
+	// The channelAddress struct contains edge information that we need to
+	// verify channel constraints in the second pass.
+	type channelAddress struct {
+		policy   *channeldb.ChannelEdgePolicy
+		capacity btcutil.Amount
+	}
+
+	chanAddrs := []*channelAddress{}
+	for i, addr := range addresses {
+		var pubkey route.Vertex
+		var channelAddr *channelAddress
+		switch a := addr.(type) {
+
+		// If a channel id is specified, fetch the channel policy and
+		// endpoints. Use the endpoints to determine the next hop
+		// pubkey.
+		case uint64:
+			info, policy1, policy2, err := r.cfg.Graph.FetchChannelEdgesByID(a)
+			if err != nil {
+				return nil, fmt.Errorf("unknown channel %v: %v", a, err)
+			}
+			from := pubkeys[i]
+			var policy *channeldb.ChannelEdgePolicy
+			switch {
+			case info.NodeKey1Bytes == from:
+				policy = policy1
+				pubkey = info.NodeKey2Bytes
+			case info.NodeKey2Bytes == from:
+				policy = policy2
+				pubkey = info.NodeKey1Bytes
+			default:
+				return nil, errDisconnectedSequence
+			}
+
+			if policy == nil {
+				return nil, errUnknownPolicy
+			}
+			channelAddr = &channelAddress{
+				policy:   policy,
+				capacity: info.Capacity,
+			}
+
+		// If a pubkey is specified, just store that pubkey. Selecting a
+		// matching channel is done in the second pass.
+		case route.Vertex:
+			pubkey = a
+		default:
+			return nil, fmt.Errorf("unknown hop address type %T",
+				addr)
+		}
+
+		pubkeys = append(pubkeys, pubkey)
+		chanAddrs = append(chanAddrs, channelAddr)
+	}
+
+	// In the second pass, select channels for hops that are specified by
+	// pubkey and verify channel policies.
+	edges := make([]*channeldb.ChannelEdgePolicy, len(addresses))
+
+	// Keep running amounts for the payment amount and the min and max for
+	// this route.
+	var (
+		amtToSend    = amt
+		minAmtToSend = lnwire.MilliSatoshi(0)
+		maxAmtToSend = lnwire.MilliSatoshi(^uint64(0))
+	)
+
+	// checkOrUpdateAmt checks or updates the running amounts while
+	// traversing the channels.
+	checkOrUpdateAmt := func(addr *channelAddress) error {
+		// If the amount to send is set, check channel constraints.
+		if amtToSend != nil {
+			// Check channel capacity.
+			capSat := lnwire.NewMSatFromSatoshis(addr.capacity)
+			if *amtToSend > capSat {
+				return fmt.Errorf("channel capacity %v "+
+					" exceeded with amt %v",
+					addr.capacity, *amtToSend)
+			}
+
+			// Check HTLC value limits.
+			if *amtToSend < addr.policy.MinHTLC ||
+				*amtToSend > addr.policy.MaxHTLC {
+
+				return errors.New("channel htlc size " +
+					"constraints violated")
+			}
+		}
+
+		// Adapt the minimum and maximum amount to send to what this
+		// channel allows.
+		//
+		// TODO(joostager): can MaxHTLC be zero?
```

Comment:
It can be 0 if the max HTLC bit isn't set. Also, the spec doesn't mention whether this value can be 0, but it should be less than min HTLC. It wouldn't make sense to have a min HTLC 0 so we can assume the lowest allowed is 1.

Reply:
I am not sure what you suggest here. To compensate for malformed channel updates? I understand the MaxHTLC=0 case, but anything else, I am not sure if we should 'try to make the best of it'.